From 9f4442453fed407d9420fa3d6d9c208980a557b3 Mon Sep 17 00:00:00 2001
From: Iaroslav Voitovych <yaroslav.voytovych@gmail.com>
Date: Fri, 9 Oct 2026 23:18:59 +0200
Subject: [PATCH BlueZ 1/4] emulator: Free hooks in btdev_destroy

Hooks added with btdev_add_hook() and not removed with btdev_del_hook()
were never freed, so LeakSanitizer reports every test that leaves a
hook installed.
---
 emulator/btdev.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/emulator/btdev.c b/emulator/btdev.c
index e00f78288..ecc24a72b 100644
--- a/emulator/btdev.c
+++ b/emulator/btdev.c
@@ -8424,6 +8424,8 @@ struct btdev *btdev_create(enum btdev_type type, uint16_t id)
 
 void btdev_destroy(struct btdev *btdev)
 {
+	int i;
+
 	if (!btdev)
 		return;
 
@@ -8438,6 +8440,9 @@ void btdev_destroy(struct btdev *btdev)
 	queue_destroy(btdev->le_per_adv, free);
 	queue_destroy(btdev->le_big, le_big_free);
 
+	for (i = 0; i < MAX_HOOK_ENTRIES; i++)
+		free(btdev->hook_list[i]);
+
 	free(btdev);
 }
 

base-commit: 7428ca2df9356363079db0f48412c20cbdb9471f
-- 
2.43.0

